src/types/provider/mod.rs

```diff
@@ -7,91 +7,20 @@
 #[cfg(test)]
 mod test;
 
-use std::{
-    convert::TryFrom,
-    fmt::{self, Display, Formatter},
-};
-
-use serde::{
-    de::{Deserialize, Deserializer, Error as SerdeError, Unexpected},
-    ser::{Serialize, Serializer},
-};
+use serde::de::Error;
 
 use types::error::{AppError, AppErrorKind};
 
-/// Identifies the underlying email provider.
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
-pub enum Provider {
-    Mock,
-    Sendgrid,
-    Ses,
-    Smtp,
-    SocketLabs,
-}
-
-impl AsRef<str> for Provider {
-    /// Return the provider as a string slice.
-    fn as_ref(&self) -> &str {
-        match *self {
-            Provider::Mock => "mock",
-            Provider::Sendgrid => "sendgrid",
-            Provider::Ses => "ses",
-            Provider::Smtp => "smtp",
-            Provider::SocketLabs => "socketlabs",
-        }
-    }
-}
+enum_boilerplate!(Provider ("env", InvalidPayload) {
```

Comment:
It's a little bit spooky-action-at-a-distance, but the reason for `InvalidPayload` here is because this error will probably come from a request payload if it does occur. It can come from settings too though and we could definitely do a better job of modelling that duality, e.g. with an `Option` member property or something. I just went with the easiest thing to code for now.

`InvalidEnv` never comes from a request payload, hence it was an easy decision to give it its own error type.

I might rethink all this on Monday...